### PR TITLE
mach: math lib adding functionality and tests

### PR DESCRIPTION
**Pull request by RokKos**
I expanded Math Lib with a few functions and tests.
I added:
- `rotateX(angle_radians: f32) Mat4x4` and 
- `rotateY(angle_radians: f32) Mat4x4`, both of them follow similiar style as existing `rotateZ`
- `set_2d_matrix(data: []const f32) Mat3x3` and 
- `set_3d_matrix(data: []const f32) Mat4x4` to set matrix to whatever values you have in 9 and 16 long arrays.  This function is similar to [wgpu-matrix mat4.set.](https://wgpu-matrix.org/docs/functions/mat4.set.html)
- `equals(a: anytype, b: @TypeOf(a)) bool` and 
- `equalsApproximately(a: anytype, b: @TypeOf(a), tolerance: f32) bool ` to easier test if two matrices are equal or close to equal. This was helpful when testing outcomes of other operations in tests. wgpu-matrix library has the same two methods [mat4.equals](https://wgpu-matrix.org/docs/functions/mat4.equals.html) and [mat4.equalsApproximately](https://wgpu-matrix.org/docs/functions/mat4.equalsApproximately.html) and zmath has similar function [isNearEqual](https://github.com/michal-z/zig-gamedev/blob/6ff65ed35758837d6f8bb38a36ae896a779e0a19/libs/zmath/src/zmath.zig#L518)

I would appreciate feedback on these new methods and I am willing to improve any of those.

That being said I would also like to contribute more to this math lib and add functionalities that I will also need in my renderer. Maybe we could have a discussion on Discord on how you see this math library. I like that is very opinionated and I would like to help it keep the same vision.

Functions that I will definitely need and implement in the future are:
- matrix transpose
- look at
- perspective matrix
- quaternions

If you have any good resources or existing implementation on how to implement those in an efficient way I would appreciate it.
Currently I was looking at [zmath](https://github.com/michal-z/zig-gamedev/tree/main/libs/zmath) and [wgpu-matrix](https://github.com/greggman/wgpu-matrix)

- [x] By selecting this checkbox, I agree to license my contributions to this project under the license(s) described in the LICENSE file, and I have the right to do so or have received permission to do so by an employer or client I am producing work for whom has this right.

### CODE REVIEW

**Comment by RokKos** [comment]
Added `cross(a: Vec3, b: Vec3) Vec3` for vector cross product and relevant tests.
It only functions on Vec3, because you can do cross-product only in 3 dimensions.

The equation was taken from [3D Math Primer for Graphics and Game Development - 2.12 Chapter Vector Cross Product](https://gamemath.com/book/vectors.html#cross_product). 

Implementation details were inspired by [zmath cross3d](https://github.com/michal-z/zig-gamedev/blob/6ff65ed35758837d6f8bb38a36ae896a779e0a19/libs/zmath/src/zmath.zig#L1948) but there Vec4 is used.

**Comment by razcore-rad** [comment]
Nitpick, shouldn't `*et_2d_matrix()` be called `*et2dMatrix()`? To keep consistent naming.

**Comment by RokKos** [comment]
> Nitpick, shouldn't `*et_2d_matrix()` be called `*et2dMatrix()`? To keep consistent naming.

Yes you are absolutely correct @razcore-rad. This slipped through

File: src/math.zig [hunk 1]
```diff
@@ -324,3 +786,422 @@ pub const mat = struct {
...
+    };
+    for (zero_value_indexes) |index| {
+        try expectEqual(scale_mat[index], 0);
+    }
+    try expectEqual(scale_mat[4 * 3 + 3], 1);
+}
+
+const degreesToRadians = std.math.degreesToRadians;
+
+// NOTE: Maybe reconsider based on feedback to join all test for rotation into one test as only
```
Comment (slimsag) [inline src/math.zig:951]:
```suggestion
// TODO: Maybe reconsider based on feedback to join all test for rotation into one test as only
```

File: src/math.zig [hunk 2]
```diff
@@ -324,3 +786,422 @@ pub const mat = struct {
...
+        const T = mat.translate3d(t);
+        const s = Vec3{ 3, 1, -5 };
+        const S = mat.scale3d(s);
+        const r = Vec3{ 30, -40, 235 };
+        const R_x = mat.rotateX(degreesToRadians(f32, r[0]));
+        const R_y = mat.rotateY(degreesToRadians(f32, r[1]));
+        const R_z = mat.rotateZ(degreesToRadians(f32, r[2]));
+
+        const R_yz = mat.mul(R_y, R_z);
+        // NOTE: This values are calculated by hand with help of matrix calculator: https://matrix.reshish.com/multCalculation.php
```
Comment (slimsag) [inline src/math.zig:1170]:
```suggestion
        // These values are calculated by hand with help of matrix calculator: https://matrix.reshish.com/multCalculation.php
```

File: src/math.zig [hunk 3]
```diff
@@ -128,8 +131,350 @@ pub const vec = struct {
...
+        const v = Vec4{ 10, 0.5, -3, -0.2 };
+        const normalized = vec.normalize(v, near_zero_value);
+        const result = Vec4{ 0.9565546486012808204, 0.04782773243006404102, -0.28696639458038424612, -0.01913109297202561641 };
+        try expectApproxEqAbs(normalized[0], result[0], 1e-7);
+        try expectApproxEqAbs(normalized[1], result[1], 1e-7);
+        try expectApproxEqAbs(normalized[2], result[2], 1e-7);
+        try expectApproxEqAbs(normalized[3], result[3], 1e-7);
+    }
+
+    //NOTE: This test ensures that zero vector is also normalized to zero vector with help of divisor
```
Comment (slimsag) [inline src/math.zig:245]:
```suggestion
    // This test ensures that zero vector is also normalized to zero vector with help of divisor
```

File: src/math.zig [hunk 4]
```diff
@@ -128,8 +131,350 @@ pub const vec = struct {
...
+    {
+        const v = Vec2{ 0, 0 };
+        const normalized = vec.normalize(v, near_zero_value);
+        try expect(normalized[0] == 0 and normalized[1] == 0);
+    }
+
+    //NOTE: This test should work but I am getting error:
```
Comment (slimsag) [inline src/math.zig:252]:
```suggestion
    // TODO: This test should work, but:
```

File: src/math.zig [hunk 5]
```diff
@@ -128,8 +131,350 @@ pub const vec = struct {
...
+    //     const normalized = vec.normalize(v, 0);
+    //     const NaN = std.math.nan(f32);
+    //     try std.testing.expectEqual(-NaN, normalized[0]);
+    //     try std.testing.expectEqual(-NaN, normalized[1]);
+
+    //     // try std.testing.expectApproxEqAbs(normalized[0], -NaN, 0.00000001);
+    //     // try std.testing.expectApproxEqAbs(normalized[1], -NaN, 0.00000001);
+    // }
+
+    //NOTE: This two test show how for small values if we add divisor we get different normalized vectors
```
Comment (slimsag) [inline src/math.zig:265]:
```suggestion
    // These two tests show how for small values if we add divisor we get different normalized vectors
```

File: src/math.zig [hunk 6]
```diff
@@ -311,6 +675,104 @@ pub const mat = struct {
         } else @compileError("Expected matrix, found '" ++ @typeName(@TypeOf(a)) ++ "'");
     }
 
+    /// Check if two matrices are exactly equal. For approximate comparison use `equalsApproximately()`
+    pub inline fn equals(a: anytype, b: @TypeOf(a)) bool {
```
Comment (slimsag) [inline src/math.zig:679]: In general exact equality is pretty non-useful, I would prefer that we make `equal` the name for approximate equality and `equalExact` the name for exact equality.

File: src/math.zig [hunk 7]
```diff
@@ -324,3 +786,422 @@ pub const mat = struct {
...
+                try expect(identity_3x3[row * 4 + column] == value);
+
+                column += 1;
+            }
+            row += 1;
+        }
+    }
+}
+
+test "mat.orto" {
```
Comment (slimsag) [inline src/math.zig:822]: This test has typos `orto` instead of `ortho`

File: src/math.zig [hunk 8]
```diff
@@ -324,3 +786,422 @@ pub const mat = struct {
...
+
+    // Computed Values
+    try expectEqual(orto_mat[0], 0.5);
+    try expectEqual(orto_mat[4 + 1], 0.4);
+    try expectEqual(orto_mat[4 * 2 + 2], -0.01);
+    try expectEqual(orto_mat[4 * 3 + 0], 0);
+    try expectEqual(orto_mat[4 * 3 + 1], -0.2);
+    try expectEqual(orto_mat[4 * 3 + 2], -0.1);
+
+    // Constant values, which should not change but we still check for completness
```
Comment (slimsag) [inline src/math.zig:833]: Please do a find and replace for `completness` -> `completeness`

File: src/math.zig [hunk 9]
```diff
@@ -324,3 +786,422 @@ pub const mat = struct {
...
+        });
+        try expect(mat.equalsApproximately(R_xyz, expected_R_xyz, tolerance));
+
+        const SR = mat.mul(S, R_xyz);
+        const expected_SR = mat.set3dMatrix(&[_]f32{
+            -1.318155125312115,  -0.5250625666689163, 3.6443452297744985,  0,
+            1.8825206147913993,  -0.7600021571513356, -0.8460473867298382, 0,
+            -1.9283628290596182, -0.383022221559489,  -3.3170697408446915, 0,
+            0,                   0,                   0,                   1,
+        });
```
Comment (slimsag) [inline src/math.zig:1194]: I don't like the `set3dMatrix` and `set2dMatrix` APIs, please remove them. I think that writing `Mat3x3{values}` is better than `mat.set3dMatrix(&[_]f32{values})`

File: src/math.zig [hunk 10]
```diff
@@ -324,3 +786,422 @@ pub const mat = struct {
...
+        const R_xyz = mat.mul(R_x, R_yz);
+        const expected_R_xyz = mat.set3dMatrix(&[_]f32{
+            -0.439385041770705,  -0.52506256666891627986, -0.72886904595489960019, 0,
+            0.6275068715971331,  -0.76000215715133560834, 0.16920947734596765363,  0,
+            -0.6427876096865394, -0.383022221559489,      0.66341394816893832989,  0,
+            0,                   0,                       0,                       1,
+        });
+        try expect(mat.equalsApproximately(R_xyz, expected_R_xyz, tolerance));
+
+        const SR = mat.mul(S, R_xyz);
```
Comment (slimsag) [inline src/math.zig:1188]: In quite a few tests, you are using uppercase variable names `const SR`, `const R`, etc. can you make those lowercase? Uppercase is reserved for type names in Zig.

**Comment by slimsag** [comment]
Thank you @RokKos for sending this. I am happy to accept such changes, and the future ones you suggested to get the functionality you need. I have similar needs.

In general, please borrow from [wgpu-matrix](https://github.com/greggman/wgpu-matrix), and beware that zmath is following directx matrix conventions - so it's a much less useful reference for us here.

There are other fundamental changes I would like to the library soon, such as introducing precision variants for e.g. f16/f32/f64 - but I need to think about the API a bit more. Happy to discuss in Discord if you have further thoughts.

File: src/math.zig (hunk 10 above)
Comment (RokKos) [reply src/math.zig:1188]: I can do that, here I was following the scientific/mathematic naming for matrices. but I can change it to lowercase.

Comment (slimsag) [reply src/math.zig:1188]: understood - but yeah let's change to lowercase I'd feel better with that 👍

**Comment by slimsag** [comment]
@RokKos curious if you're still interested in updating this, I can take a look at making the changes and merging later tonight if not

**Comment by RokKos** [comment]
Hi @slimsag sorry for no response... I have been on last minute vacation before starting new job...
I think I can do those minor changes, but I actually don't know exactly how would I implement feature to initialize matrices with `Mat3x3 {values}`. So if you can help me with that or give some pointers that would be beneficial

**Comment by slimsag** [comment]
@RokKos no worries, the changes were simple to make so I went ahead and made them real quick. Thanks for the awesome PR! I appreciate it!

**Comment by RokKos** [comment]
Perfect thank you @slimsag !!!

**Pull request merged**